### PR TITLE
kill unresponsive servers on quit

### PR DESCRIPTION
**Pull request by dyfer**
## Purpose and Motivation

`Server -quit` does not quit unresponsive servers. This PR adds `kill` command to the server quit process if the given server is unresponsive.

IMO this should prevent majority of issues with hanging servers, since we'd be killing them e.g. on exit and recompile. However, I realize that this is also quite a strong measure and I'd like to hear if anyone feels like this has any major drawbacks.

## Types of changes

- Bug fix
- New feature

## To-do list

- [x] Code is tested
~~- [ ] All tests are passing~~
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

**Comment by brianlheim** [comment]
i think this is too strong an action to take. there may be legitimate reasons for the server to be unresponsive, such as loading a large file, and the gracious thing to do in that case is leave it up in case it becomes responsive again. IMHO, better to leave that severe of a consequence to the user's discretion.

**Comment by dyfer** [comment]
I don't think I understand. If the user wants to quit the server, what would be the justified reason for the server to keep running?

A very common server hang scenario in my experience is when the soundcard gets disconnected while the server is running (e.g. SC running in the background) and calling `server.quit` or recompiling library does not solve that. This PR would solve this.

**Comment by brianlheim** [comment]
ok good point, i didn't consider this very thoroughly. in that case, i have a different review. sorry for the noise!

File: SCClassLibrary/Common/Control/Server.sc [hunk 1]
```diff
@@ -1085,9 +1085,10 @@ Server {
 
 		addr.sendMsg("/quit");
 
-		if(watchShutDown and: { this.unresponsive }) {
+		if(this.unresponsive) {
 			"Server '%' was unresponsive. Quitting anyway.".format(name).postln;
-			watchShutDown = false;
+			pid !? {thisProcess.platform.killProcessByID(pid)};
```
Comment (brianlheim) [inline SCClassLibrary/Common/Control/Server.sc:1090]: i think this should be turned into an `if` using pid.notNil as a condition; in the true branch kill by PID and post a more forceful message (like "Server '%' was unresponsive. Killing forcefully.") and in the false branch post the original message.

File: SCClassLibrary/Common/Control/Server.sc [hunk 2]
```diff
@@ -1085,9 +1085,10 @@ Server {
...
+			if(watchShutDown) {watchShutDown = false};
```
Comment (brianlheim) [inline SCClassLibrary/Common/Control/Server.sc:1091]: this line can be left as is; watchShutDown is either true or false so this change doesn't alter the behavior in any way.

**Comment by dyfer** [comment]
@brianlheim thanks for the review, I made the requested changes

**Comment by dyfer** [comment]
When working on documentation, I realized that `-reboot` is not calling `quit` on unresponsive server. I'd like to propose removing that, it is unclear to me what the use case for this would be. Shouldn't we clean up when the server is unresponsive? I'll keep this as a separate commit so I can easily change that during the review.

**Comment by dyfer** [comment]
Documentation updated.

**Comment by dyfer** [comment]
I think this is minor but previously we were not posting `"Server 'xxx' was unresponsive. Quitting anyway."` when `watchShutDown = false`. I now added a check to maintain previous posting...

**Comment by dyfer** [comment]
Thanks again for the review @brianlheim. Since you mentioned it would be good for someone else to take a look before merging this I wonder if @jamshark70 or @joshpar could take a look?

File: SCClassLibrary/Common/Control/Server.sc [hunk 3]
```diff
@@ -1085,8 +1085,15 @@ Server {
 
 		addr.sendMsg("/quit");
 
-		if(watchShutDown and: { this.unresponsive }) {
-			"Server '%' was unresponsive. Quitting anyway.".format(name).postln;
+		if(this.unresponsive) {
+			if(pid.notNil) {
+				"Server '%' was unresponsive. Killing forcefully.".format(name).postln;
```
Comment (telephon) [inline SCClassLibrary/Common/Control/Server.sc:1090]: As we have become more sensitive to human metaphors recently, a suggestion that has less police associations:

```
"Server '%' is currently unresponsive. Forcing process to stop via system command."
```

**Comment by elgiano** [comment]
Looks good to me too! I agree with telephon's comment above.

**Comment by dyfer** [comment]
Thanks @telephon I updated the message.
@brianlheim do you think we need more discussion on this?

**Comment by brianlheim** [comment]
@dyfer nope, i think it's good! thanks for the reviews and updates everyone.

**Comment by adcxyz** [comment]
> Turns out that `-boot` method actually calls `quit` if the server is unresponsive. The boot sequence in sclang is a bit messy :)

:-)
FYI, I had a full cleanup proposal for this that worked well, but ended up not being merged: 
https://github.com/adcxyz/supercollider/tree/server_ServerProcess
I think the solutions implemented there were a good result capturing the insights 
of a thorough discussion process, so it would be worth redoing that branch 
based on the today's state of the server code.

**Pull request merged**